packages/modeling/src/operations/booleans/martinez/connect_edges.js

```diff
@@ -0,0 +1,187 @@
+/*
+ * Implementation of the Martinez 2D polygon clipping algorithm
+ * Copyright (c) 2018 Alexander Milevski
+ * https://github.com/w8r/martinez
+ */
+
+import compareEvents from './compare_events.js'
+import Contour from './contour.js'
+
+/**
+ * @param  {Array.<SweepEvent>} sortedEvents
+ * @return {Array.<SweepEvent>}
+ */
+const orderEvents = (sortedEvents) => {
+  let event, i, len, tmp
+  const resultEvents = []
+  for (i = 0, len = sortedEvents.length; i < len; i++) {
+    event = sortedEvents[i]
+    if ((event.left && event.inResult) ||
+      (!event.left && event.otherEvent.inResult)) {
+      resultEvents.push(event)
+    }
+  }
+  // Due to overlapping edges the resultEvents array can be not wholly sorted
+  let sorted = false
+  while (!sorted) {
+    sorted = true
+    for (i = 0, len = resultEvents.length; i < len; i++) {
+      if ((i + 1) < len &&
+        compareEvents(resultEvents[i], resultEvents[i + 1]) === 1) {
+        tmp = resultEvents[i]
+        resultEvents[i] = resultEvents[i + 1]
+        resultEvents[i + 1] = tmp
+        sorted = false
+      }
+    }
+  }
+
+  for (i = 0, len = resultEvents.length; i < len; i++) {
+    event = resultEvents[i]
+    event.otherPos = i
+  }
+
+  // imagine, the right event is found in the beginning of the queue,
+  // when his left counterpart is not marked yet
+  for (i = 0, len = resultEvents.length; i < len; i++) {
+    event = resultEvents[i]
+    if (!event.left) {
+      tmp = event.otherPos
+      event.otherPos = event.otherEvent.otherPos
+      event.otherEvent.otherPos = tmp
+    }
+  }
+
+  return resultEvents
+}
+
+/**
+ * @param  {Number} pos
+ * @param  {Array.<SweepEvent>} resultEvents
+ * @param  {Object>}    processed
+ * @return {Number}
+ */
+const nextPos = (pos, resultEvents, processed, origPos) => {
+  let newPos = pos + 1
+  const p = resultEvents[pos].point
+  let p1
```

Comment:
I agree that the code is structured in an unclear way, but I think it works. It is the original logic from martinez.

The reason it works is because `p1` is initialized if `newPos < length`, but the same condition exists on the while loop. So it won't enter the loop unless `p1` is initialized.